accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,245 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+**definitely assigned**.
+
+It is not a compile time error to write to a `final` local variable if that
+variable is **definitely unassigned** (and hence does not have an initializer).
+
+It is a compile time error to read a local variable when the variable is
+**definitely unassigned** unless the variable is non-final and has nullable
+type, either explicitly provided or inferred directly from the initalizer.  This
+includes variables marked `late` and/or `final`.
+
+The erors specified above are summarized in the following table.  A variable
+which has an initializer (explicit or implicit) is always considered definitely
+assigned, and is never considered definitely unassigned.
+
+
+Read Errors:
+
+| Declaration form  | Def. Assigned | Not Def. Assigned | Def. Unassigned |
+| -------- ---------| ------------- | ----------------- | --------------- |
+| var x;            | Ok            | Error             | Error           |
+| final x;          | Ok            | Error             | Error           |
+| int x;            | Ok            | Error             | Error           |
+| int? x;           | Ok            | Ok                | Ok              |
+| final int x;      | Ok            | Error             | Error           |
+| final int? x;     | Ok            | Error             | Error           |
+| late var x;       | Ok            | Ok                | Error           |
+| late final x;     | Ok            | Ok                | Error           |
+| late T x;         | Ok            | Ok                | Error           |
+| late final T x;   | Ok            | Ok                | Error           |
+
+Write Errors:
+
+| Declaration form  | Def. Assigned | Not Def. UnAssigned | Def. Unassigned |
+| -------- ---------| ------------- | ------------------- | --------------- |
+| var x;            | Ok            | Ok                  | Ok              |
+| final x;          | Error         | Error               | Ok              |
+| int x;            | Ok            | Ok                  | Ok              |
+| int? x;           | Ok            | Ok                  | Ok              |
+| final int x;      | Error         | Error               | Ok              |
+| final int? x;     | Error         | Error               | Ok              |
+| late var x;       | Ok            | Ok                  | Ok              |
+| late final x;     | Error         | Ok                  | Ok              |
+| late T x;         | Ok            | Ok                  | Ok              |
+| late final T x;   | Error         | Ok                  | Ok              |
+
+### Local variables and initialization based inference
+
+Local variables with explicitly written types are given the declared types as
+written.  The declared type of the variable is considered a "type of interest"
+in the sense defined in the flow analysis specification.  If the variable has an
+initializer (explicit or implicit) and is not `final`, then the declaration is
```

Comment:
@munificent @lrhn @eernstg @johnniwinther should we maintain the restriction to final here, or do the promotion for all variables?  @munificent argues that it is more uniform to treat all variables the same here.  My reasoning was that for a final variable, it is useless to declare it as nullable and then assign a non-nullable variable, since you can never re-assign it to null: you should just declare it with the non-nullable version.  But I don't feel that strongly about it.

Reply:
@eernstg 

> How about `Null`? Do we actually want `int? i = null;` to promote `i` to `Null` immediately, and even `int? i;`? It's kind of nice that we prevent method invocations that would fail because the value is definitely null, but I suspect that it would be impractical in many other ways, e.g., because `int` isn't a subtype of `Null` so we can't proceed to promote to `int`.

I'm not sure what you're asking here.  As currently specified and implemented, if `int?` is a type of interest, then `int` is also type of interest, but `Null` is not.  Are you proposing that we make it so?

> So it does make sense to promote the type of a final variable even though it doesn't get a new value. So I'd prefer to include final variables here.

I'm don't really see the connection between these two.  The argument against promoting the final variable on initialization is not that it can't be promoted, but that it can never be demoted, so declaring the looser type is pointless.




> For the point about being nullable: The static type of the initializing expression (or of the right hand side of an initializing assignment) may be nullable, and it still makes sense to test and detect that the value is non-null, even though it can't get a new value.

I don't understand what you're trying to say here, can you give an example?

Reply:
> Maybe define 'typed' and 'untyped' variables first, to make it a bit more explicit that we have text about the former first, and further down we'll have text about the latter—maybe we could even have a subsubsubtitle saying "Typed Variables" and "Untyped Variables".

I'm not convinced this will make things clearer.  Defining what a "typed variable" is explicitly is a bit painful, unless define it as "not an untyped variable".  The latter is clear and precise, but requires untyped variables to be introduced first.  Given that I don't actually need a term for "typed variable", I'm not sure why it helps things to introduce it as a concept.

Reply:
>> How about `Null`? ..

> I'm not sure what you're asking here. As currently specified and
> implemented, if `int?` is a type of interest, then `int` is also type of
> interest, but `Null` is not. Are you proposing that we make it so?

My reading of the following suggests that the variable should be promoted to have type `Null` immediately at the declaration:

> If the variable has an initializer (explicit or implicit) and is not `final`,
> then the declaration is treated as an assignment for the purposes
> of promotion.

This would occur even for `int? x;`, where the initializer is an implicit `null`. But are you saying that `Null` is not a type of interest even in the case where the variable has been promoted to have type `Null`?:

>  if `int?` is a type of interest, then `int` is also type of interest, but `Null` is not.

In any case, my point was that the variable was promoted to have type `Null`, not whether `Null` is a type of interest. I was just noting that, apparently, variables of nullable type with no non-null initializer would have type `Null` until they are demoted to some other type. 

This might actually be fine, but it seemed somewhat surprising to me. For instance, it could be confusing for developers if they get "`Null` does not have a member `isEven`" rather than "`isEven` cannot be invoked when the receiver has the nullable type `int?`".

----

> declaring the looser type is pointless.

Right, I was just thinking that "pointless" is usually serious enough for a lint, but not for giving a situation a special treatment in the language. So I thought it would make sense to perform the promotion of a final variable based on its initializer (that is: remove the exception for `final`), and then leave it to a lint to tell the developer that the declared type is too loose and will basically be ignored.

>> For the point about being nullable ..

Same topic, no need to go deeper with that.

----

>> Maybe define 'typed' and 'untyped' variables first

This was just a readability issue. The text talks about typed variables for 3 paragraphs and a code snippet, and then about untyped variables basically to the end of the section (which is more like multiple "pages"). I thought it would be helpful to make this overall structure visible in the presentation, rather than tucking away the definition of 'untyped variable' at the end of the 5th paragraph. There's no need to make 'typed variable' anything other than "not 'untyped variable'". The section could start with this:

```
An **untyped variable** is a variable declared as `var x;`, `final x;`, `late var x;`,
or `late final x;`, and a **typed variable** is a variable which is not untyped.
We discuss typed variables first, and then untyped ones.
```

Reply:
> > > How about `Null`? ..
> 
> > I'm not sure what you're asking here. As currently specified and
> > implemented, if `int?` is a type of interest, then `int` is also type of
> > interest, but `Null` is not. Are you proposing that we make it so?
> 
> My reading of the following suggests that the variable should be promoted to have type `Null` immediately at the declaration:

@eernstg I may be missing something, but I'm not following you at all.  Let me try to walk through this as I understand it, and you can tell me where you're seeing something different.

> 
> > If the variable has an initializer (explicit or implicit) and is not `final`,
> > then the declaration is treated as an assignment for the purposes
> > of promotion.
> 
> This would occur even for `int? x;`, where the initializer is an implicit `null`. But are you saying that `Null` is not a type of interest even in the case where the variable has been promoted to have type `Null`?:

I don't understand where you're getting to the point that the variable has been promoted to `Null`.  We do not promote on all assignments - only assignments where the RHS has a type of interest.

> 
> > if `int?` is a type of interest, then `int` is also type of interest, but `Null` is not.
> 
> In any case, my point was that the variable was promoted to have type `Null`, not whether `Null` is a type of interest. I was just noting that, apparently, variables of nullable type with no non-null initializer would have type `Null` until they are demoted to some other type.

Right.  So where did the variable get promoted?  The sentences you quite imply that `int? x = null` makes `int` and `int?` a type of interest, and treats `int? x = null` as if it were `int? x; x = null` for the purposes of promotion.  The assignment `x = null` causes the question to be asked "is `Null` a type of interest".  The answer is "no", so no promotion happens.  So again, where are you seeing a promotion to `Null` happening?

Reply:
Right, I thought that `int? x = null` would make `= null` an initializing assignment, but I can see that this is explicitly restricted to untyped valiables. So it's a regular assignment after all, so I agree that it won't promote.

This leaves only one topic alive in this thread: 

> should we maintain the restriction to final